browser/updater/updater_p3a_unittest.cc

```diff
@@ -0,0 +1,134 @@
+
+namespace brave_updater {
+
+class UpdaterP3ATest : public ::testing::TestWithParam<bool> {
+ public:
+  UpdaterP3ATest() = default;
+  ~UpdaterP3ATest() override = default;
+
+  void SetUp() override {
+    ResetHistogramTester();
+    RegisterLocalStatePrefs(local_state_.registry());
+  }
+
+ protected:
+  base::test::TaskEnvironment task_environment_{
+      base::test::TaskEnvironment::TimeSource::MOCK_TIME};
+  TestingPrefServiceSimple local_state_;
+  std::unique_ptr<base::HistogramTester> histogram_tester_;
+
+  bool IsUsingOmaha4() const { return GetParam(); }
+
+  void SimulateLaunch(int advance_by_days, std::string current_version) {
+    task_environment_.AdvanceClock(base::Days(advance_by_days));
+    ReportLaunch(base::Time::Now(), current_version, IsUsingOmaha4(),
+                 &local_state_);
+  }
+
+  void ResetHistogramTester() {
+    histogram_tester_ = std::make_unique<base::HistogramTester>();
+  }
+
+  UpdateStatus StatusUpdate() {
+    return IsUsingOmaha4() ? UpdateStatus::kUpdatedWithOmaha4
+                           : UpdateStatus::kUpdatedWithLegacy;
+  }
+
+  UpdateStatus StatusNoUpdate() {
+    return IsUsingOmaha4() ? UpdateStatus::kNoUpdateWithOmaha4
+                           : UpdateStatus::kNoUpdateWithLegacy;
+  }
+};
+
+TEST_P(UpdaterP3ATest, TestNoReportInInitialLaunch) {
+  SimulateLaunch(0, "1.0.0.0");
+  histogram_tester_->ExpectTotalCount(kUpdateStatusHistogramName, 0);
+}
+
+TEST_P(UpdaterP3ATest, TestReportNoUpdateIfNoVersionChange) {
+  SimulateLaunch(0, "1.0.0.0");
+  SimulateLaunch(1, "1.0.0.0");
+  histogram_tester_->ExpectUniqueSample(kUpdateStatusHistogramName,
+                                        StatusNoUpdate(), 1);
+}
+
+TEST_P(UpdaterP3ATest, TestReportsUpdateIfPatchVersionChanges) {
+  SimulateLaunch(0, "1.0.0.0");
+  SimulateLaunch(1, "1.0.0.1");
+  histogram_tester_->ExpectUniqueSample(kUpdateStatusHistogramName,
+                                        StatusUpdate(), 1);
+}
+
+TEST_P(UpdaterP3ATest, TestReportsUpdateIfBuildVersionChanges) {
+  SimulateLaunch(0, "1.0.0.0");
+  SimulateLaunch(1, "1.0.1.0");
+  histogram_tester_->ExpectUniqueSample(kUpdateStatusHistogramName,
+                                        StatusUpdate(), 1);
+}
+
+TEST_P(UpdaterP3ATest, TestReportsUpdateIfMinorVersionChanges) {
+  SimulateLaunch(0, "2.0.0.0");
+  SimulateLaunch(1, "2.1.0.0");
+  histogram_tester_->ExpectUniqueSample(kUpdateStatusHistogramName,
+                                        StatusUpdate(), 1);
+}
+
+TEST_P(UpdaterP3ATest, TestReportsUpdateIfMajorVersionChanges) {
+  SimulateLaunch(0, "2.0.0.0");
+  SimulateLaunch(1, "3.0.0.0");
+  histogram_tester_->ExpectUniqueSample(kUpdateStatusHistogramName,
+                                        StatusUpdate(), 1);
+}
+
+TEST_P(UpdaterP3ATest, TestReportsUpdateIfVersionDowngrade) {
+  SimulateLaunch(0, "2.0.0.0");
+  SimulateLaunch(1, "1.0.0.0");
+  histogram_tester_->ExpectUniqueSample(kUpdateStatusHistogramName,
```

Comment:
not sure if this is right? Should we ignore downgrades since those obviously weren't done by the updater?